### DISCUSSION TITLE
Reduce/spread CPU clock on object detection on multiple cores

### DISCUSSION THREAD

**Discussion by aafaqin** [ultralytics/ultralytics#12187] [Q&A]
I am trying to run tensorrt inference on yolov8x, mu GPU utilisation and GPU memory are less than 25% utilized.
![image](https://github.com/ultralytics/ultralytics/assets/22724087/45cd7ef5-dc9f-4e77-94c6-9be4278356ce)

Few of my cores are more than 90% utilized. But half of my cores are just idle
![image](https://github.com/ultralytics/ultralytics/assets/22724087/aff5d44d-e9c7-4c43-8cfc-0ddd506a313a)
I am using the standard predict function with tensorrt.
Nothing fancy.
I can buy better CPU also but i just want to improve my FPS nothing else required

**Comment by pderrenger** [comment]
It seems like you're experiencing an imbalance in CPU core utilization during your TensorRT inference with YOLOv8x. Here are a couple of suggestions that might help improve the distribution across cores and possibly enhance the FPS:

1. **Set Thread Affinity:** You can try setting the thread affinity manually to spread the workload more evenly across all available cores. This is done via the OS, and the method varies between Windows, Linux, and macOS.

2. **Increase Batch Size:** If your GPU utilization is low, increasing the batch size for your inferences can help utilize your GPU more effectively, which might reduce the load on your CPU.

3. **Optimize TensorRT Engine:** When building your TensorRT engine, ensure you're maximizing the usage of available GPU resources by enabling appropriate optimizations such as layer fusion, precision calibration (INT8/FP16), and more.

Here's a small example of how you might configure your TensorRT to utilize FP16 optimization if not already set:
```python
from ultralytics import YOLO

model = YOLO('yolov8x.pt')
model.export(format='engine', half=True)  # Export with FP16
```

Also, if you're open to upgrading hardware, increasing your CPU's core and thread count might not necessarily resolve the imbalance but could improve overall throughput.

If none of these changes help, it might be worth profiling your application to see if there are specific bottlenecks in your processing pipeline that are causing the disproportionate CPU load.
